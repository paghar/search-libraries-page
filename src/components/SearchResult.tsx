import SelectBox from "./misc/SelectBox";
import {sortItems} from "@/utils/constants/selectBox";
import {SearchAction} from "@/utils/enum/search";
import {InitialStateProps} from "@/utils/interface/search";

interface IProps{   
  searchInfos:InitialStateProps,  
  dispatch:any, 
}

const SearchResult = (({searchInfos,dispatch}:IProps)=>{

  const items = searchInfos?.searchItems?.map((item:any,index)=>{        
    return (
      <div key={index} className="mt-4 flex">
        <div>
          <div className="flex h-8 border-l-4 border-pink-600"/>               
          <div className="flex h-24 border-l-4 border-gray-400"/>   
        </div>
        <div>
          <div className="flex h-8">
            <span className="ml-2 text-lg text-pink-600 font-bold">{item.name}</span>
          </div>
          <div className="flex flex-col py-2 h-24">
            <span className="ml-2 text-gray-700">
              {item.description}
            </span>
            <span className="ml-2 text-gray-400">
              stars:{item.stars}, rank:{item.rank}, latest_release_published_at:{item.latest_release_published_at}
            </span>
          </div>
        </div>
      </div>
    );    
  });
  
  return(     
    <>
      <div className="flex justify-between">
        <h6 className="text-l font-bold text-gray-800 tracking-widest uppercase">{searchInfos?.totalSearchItem} packages</h6>
        <SelectBox
          id="sort"         
          items={sortItems}
          selectedValue={searchInfos?.sort}
          onChange={(e) => dispatch({type: SearchAction.SetSortType,payload:e.target.value})} 
        />      
      </div>       
      <div className="space-y-12 px-2 xl:px-16 mt-12">{items}</div> 
    </>
  );
});
    
export default SearchResult;
    