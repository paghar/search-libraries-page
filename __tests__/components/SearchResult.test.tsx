import React from "react";
import {render} from "@testing-library/react";
import SearchResult from "../../src/components/SearchResult";

const searchInfos = {
  searchText:"npm",
  sort: "stars",     
  searchItems: [
    {name: "Item 1", description: "Description 1", stars: 10, rank: 1, latest_release_published_at: "2022-01-01"},
    {name: "Item 2", description: "Description 2", stars: 20, rank: 2, latest_release_published_at: "2022-01-02"},
    {name: "Item 3", description: "Description 3", stars: 30, rank: 3, latest_release_published_at: "2022-01-03"}
  ],
  totalSearchItem: "3",
  pageNumber:2
    
};
const dispatch = jest.fn();

describe("Button Component", () => {
  const {container} = render(
    <SearchResult searchInfos={searchInfos} dispatch={dispatch} />
  );   
  it("Renders Snapshot Buttons", () => {    
    expect(container).toMatchSnapshot();
  });
    
});

describe("SearchResult Component", () => {
  test("renders search items and handles sorting selection", () => {  

    const {getByText} = render(
      <SearchResult searchInfos={searchInfos} dispatch={dispatch} />
    );

    // Check if search items are rendered correctly
    searchInfos.searchItems.forEach(item => {
      expect(getByText(item.name)).toBeInTheDocument();
      expect(getByText(item.description)).toBeInTheDocument();
    });

    // Check if total search items are displayed
    expect(getByText(`${searchInfos.totalSearchItem} packages`)).toBeInTheDocument();

  });
});
