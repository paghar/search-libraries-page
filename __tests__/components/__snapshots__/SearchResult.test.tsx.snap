// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Renders Snapshot Buttons 1`] = `
<div>
  <div
    class="2xl:text-base xl:text-base lg:text-base md:text-sm sm:text-xs"
  >
    <div
      class="flex justify-between flex-wrap"
    >
      <div
        class=" font-bold text-gray-800 tracking-widest uppercase"
      >
        3
         packages
      </div>
      <div>
        <div
          class="relative inline-flex self-center"
        >
          <svg
            class="text-white bg-pink-700 absolute top-0 right-0 m-2 pointer-events-none p-2 rounded"
            height="25px"
            version="1.1"
            viewBox="0 0 38 22"
            width="25px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
              id="ZahnhelferDE—Design"
              stroke="none"
              stroke-width="1"
            >
              <g
                fill="#ffffff"
                fill-rule="nonzero"
                id="ZahnhelferDE–Icon&Asset-Download"
                transform="translate(-539.000000, -199.000000)"
              >
                <g
                  id="Icon-/-ArrowRight-Copy-2"
                  transform="translate(538.000000, 183.521208)"
                >
                  <polygon
                    id="Path-Copy"
                    points="33 5.38477631 33 31.3847763 29 31.3847763 28.999 9.38379168 7 9.38477631 7 5.38477631"
                    transform="translate(20.000000, 18.384776) rotate(135.000000) translate(-20.000000, -18.384776) "
                  />
                </g>
              </g>
            </g>
          </svg>
          <select
            class="text-l font-bold rounded border-2 border-pink-700 text-gray-600 h-10 w-60 pl-5 pr-10 bg-white hover:border-gray-400 focus:outline-none appearance-none"
            id="sort"
          >
            <option
              value="1"
            >
              Stars
            </option>
            <option
              value="2"
            >
              Ranks
            </option>
            <option
              value="3"
            >
              Dependents
            </option>
            <option
              value="4"
            >
              Latest Release
            </option>
            <option
              value="5"
            >
              Contributions
            </option>
            <option
              value="6"
            >
              Newest
            </option>
          </select>
        </div>
      </div>
    </div>
    <div
      class="space-y-12 px-2 mt-12"
    >
      <div
        class="mt-4 flex"
      >
        <div>
          <div
            class="flex h-8 border-l-4 border-pink-600"
          />
          <div
            class="flex h-24 border-l-4 border-gray-400"
          />
        </div>
        <div>
          <div
            class="flex h-8"
          >
            <span
              class="ml-2 text-pink-600 font-bold"
            >
              Item 1
            </span>
          </div>
          <div
            class="flex flex-col py-2 h-24"
          >
            <span
              class="ml-2 text-gray-700 overflow-y-auto h:12"
            >
              Description 1
            </span>
            <span
              class="ml-2 text-gray-400"
            >
              stars:
              10
              , rank:
              1
              , latest_release_published_at:
              2022-01-01
            </span>
          </div>
        </div>
      </div>
      <div
        class="mt-4 flex"
      >
        <div>
          <div
            class="flex h-8 border-l-4 border-pink-600"
          />
          <div
            class="flex h-24 border-l-4 border-gray-400"
          />
        </div>
        <div>
          <div
            class="flex h-8"
          >
            <span
              class="ml-2 text-pink-600 font-bold"
            >
              Item 2
            </span>
          </div>
          <div
            class="flex flex-col py-2 h-24"
          >
            <span
              class="ml-2 text-gray-700 overflow-y-auto h:12"
            >
              Description 2
            </span>
            <span
              class="ml-2 text-gray-400"
            >
              stars:
              20
              , rank:
              2
              , latest_release_published_at:
              2022-01-02
            </span>
          </div>
        </div>
      </div>
      <div
        class="mt-4 flex"
      >
        <div>
          <div
            class="flex h-8 border-l-4 border-pink-600"
          />
          <div
            class="flex h-24 border-l-4 border-gray-400"
          />
        </div>
        <div>
          <div
            class="flex h-8"
          >
            <span
              class="ml-2 text-pink-600 font-bold"
            >
              Item 3
            </span>
          </div>
          <div
            class="flex flex-col py-2 h-24"
          >
            <span
              class="ml-2 text-gray-700 overflow-y-auto h:12"
            >
              Description 3
            </span>
            <span
              class="ml-2 text-gray-400"
            >
              stars:
              30
              , rank:
              3
              , latest_release_published_at:
              2022-01-03
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
